src/raft.rs

```diff
@@ -1225,7 +1230,10 @@ impl<T: Storage> Raft<T> {
         if pr.state == ProgressState::Replicate && pr.ins.full() {
             pr.ins.free_first_one();
         }
-        if pr.matched < self.raft_log.last_index() {
+        if pr.matched < self.raft_log.last_index()
+            || self.raft_log.term(pr.next_idx - 1).is_err() // Does it request snapshot?
```

Comment:
How about `pr.next_idx < pr.matched`? Querying term and entries is expensive operations.